docs/dev/bootstrap_services.md

```diff
@@ -0,0 +1,92 @@
+#### Bootstrap Service Records ####
+
+For the purposes of diagnosing installation failures that occur during bootstrapping, the
+progresses of services running on the bootstrap machines are tracked in json files in the
+/var/log/openshift directory. The progress for each service is tracked in its own file. For
+example, the bootkube service progress is tracked in the /var/log/openshift/bootkube.json
+file. The following progress events are tracked.
+* A service adds an entry when the service starts.
+* A service adds an entry when the service ends. The entry includes the result of the service
+invocation, either success or failure. If the invocation failed, then the entry includes
+the line number of the error and the last three lines from the service's journal log.
+* A service adds an entry when a stage of the service starts. An example of a service stage
+is the cvo-bootstrap stage of the bootkube service. During that stage, the
+cluster-version-operator renders its manifests.
+* A service  adds an entry when a stage of the service ends. This is similar to the entry
+added when a service ends, including having a result and error information if applicable.
+
+##### Managing Service Records in a Service #####
+
+To track its progress, a service should source the /usr/local/bin/bootstrap-service-record.sh
+script. When a service sources the script, the script will add an entry to the json file for
+the service indicating that the service started. When the service ends, either successfully
+or due to an error, the script will add an entry to the json file for the service indicating that
+the service ended. The script will consider whether the last command executed was successful or
+not in order to determine whether the service was successful.
+
+For tracking stages, the service should call functions from the sourced script.
+* The service should call the `record_service_stage_start` function when a stage of the service
+starts. The function takes the name of the stage as its single argument.
+* The service should call the `record_service_stage_success` function when a stage of the service
+ends successfully.
+* The service should call the `record_service_stage_failure` function when a stage of the service
+ends due to a failure. The script will automatically record an entry for a stage failure if the
+service ends during the execution of a stage.
+
+###### Pre- and Post-Commands ######
+
+If a service has pre- or post-commands that could either run for significant periods or could
+potentially fail, then those commands should add to the json file as well. Such a command should
+source the same /usr/local/bin/bootstrap-service-record.sh script. It should also set either the
```

Comment:
I don't think I'm grasping the `PRE-` `POST_COMMAND` concept although this description is fine. 

What value does this add? It doesn't seem like it, but does it link services together?

Reply:
Let's take the kubelet service as an example.

As described in the doc, that service has a pre-command of `/usr/local/bin/kubelet-pause-image.sh`. What that means is that, when the kubelet service runs, the first step of the service is to run the kube-pause-image script. If the kube-pause-image script fails, then the service fails. On restart of the service, the kube-pause-image script is run again. The main kubelet command does not start until the kube-pause-image script succeeds.

When analyzing a bootstrap gather bundle, we want to be able to look at the kubelet.json file to see (1) whether the service completed successfully, (2) whether the service failed and restarted, or (3) whether the service is still running. If the service is restarting because the kube-pause-image script is failing, then we want to be able to ascertain that from `preCommand: kube-pause-image` entries in the kubelet.json file. Likewise, if the service is stuck waiting on the kube-pause-image script to complete, then we want to be able to see that from `preCommand: kube-pause-image, phase: start` entry without a subsequent `preCommand: kube-pause-image, phase: end` entry.

You could almost think of the pre- and post-commands entries as similar to stage entries except that they run before and after the main command. The entries are included in the same json file as the rest of the entries for the service.

Reply:
Here is an example of a kubelet.service file.
```
[
  {
    "timestamp": "2021-03-12T00:43:37Z",
    "preCommand": "kubelet-pause-image",
    "phase": "start"
  },
  {
    "timestamp": "2021-03-12T00:43:38Z",
    "preCommand": "kubelet-pause-image",
    "phase": "end",
    "result": "success"
  },
  {
    "timestamp": "2021-03-12T00:43:38Z",
    "phase": "start"
  }
]
```